ao-tools/ao-usbload/ao-usbload.1
.\"
.\" Copyright © 2009 Keith Packard <keithp@keithp.com>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
.\"
.\"
.TH AO-LOAD 1 "ao-usbload" ""
.SH NAME
ao-usbload \- flash a program to an ARM-based AltOS device
.SH SYNOPSIS
.B "ao-usbload"
[\-T \fItty-device\fP]
[\--tty \fItty-device\fP]
[\-D \fIaltos-device\fP]
[\--device \fIaltos-device\fP]
[\--cal \fIradio-calibration\fP]
[\--serial \fserial-number\fP]
\fIfile.elf\fP or \fIfile.ihx\fP
.SH DESCRIPTION
.I ao-usbload
loads the specified .elf or .ihx file into the target device flash
memory via the AltOS boot loader, using either existing serial number
and radio calibration values or taking either of those from the
command line.
.SH OPTIONS
.TP
\-T tty-device | --tty tty-device
This selects which tty device the debugger uses to communicate with
the target device. The special name 'BITBANG' directs ao-dbg to use
the cp2103 connection, otherwise this should be a usb serial port
connected to a suitable cc1111 debug node.
.TP
\-D AltOS-device | --device AltOS-device
Search for a connected device. This requires an argument of one of the
following forms:
.IP
TeleMega:2
.br
TeleMega
.br
2
.IP
Leaving out the product name will cause the tool to select a suitable
product, leaving out the serial number will cause the tool to match
one of the available devices.
.TP
\-s serial-number | --serial serial-number
This programs the device serial number into the image. If no serial
number is specified, then the existing serial number, if any, will be
read from the device.
.TP
\-c radio-calibration | --cal radio-calibration This programs the
radio calibration value into the image for hardware which doesn't have
any eeprom storage for this value. If no calibration value is
specified, an existing calibration value will be used. The value here
can be computed given the current radio calibration value, the
measured frequency and the desired frequency:
.IP
	cal' = cal * (desired/measured)
.IP
The default calibration value is 7119667.
.SH USAGE
.I ao-usbload
reads the specified .elf file into memory, edits the image to
customize it using the specified serial number and radio calibration
values. It then connects to the debug dongle and writes the program to
the target device flash memory.
.SH AUTHOR
Keith Packard


ChangeLog
commit 0cd443d5e55b2c3b97ecf53389ff76bc4bc6018f
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Sep 16 12:38:33 2012 -0700

    Add release notes for version 1.1.1
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit cb48dd0b03b445437f751028e8383610b65b0a68
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Sep 16 02:39:25 2012 -0700

    altos: Fix telemini sdcdb init file
    
    Set the driver source path for the debugger
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 3520bbf1ed6461d1ce7af001c529563a3cffa3c9
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Sep 16 00:58:20 2012 -0700

    altosui: Gather Tm sensor data in Monitor Idle mode
    
    Tm is pretty much the same as TM for the analog sensors, it's just
    missing the accelerometr. Use the same code for constructing an
    AltosRecord for it
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 31f5a02654cbf172beed25f4c518dfb7be8c714e
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Sep 14 16:56:04 2012 -0700

    altos: Need to use 16-bit counts for ao_xmem functions
    
    Trying to use 8-bit counts is a nice optimization which fails when the
    count is larger than 255, as is the case with clearing the flash block
    in the AT45 driver. This bug resulted in the inability to erase
    flights on TeleMetrum v1.0 boards.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit b898cf0a2abf2b0478d5afc5aca030c6b4c8bd0b
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Sep 14 12:59:31 2012 -0700

    altosui: Show over-ground-distance in Descent tab
    
    Helps to know where the rocket might land.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit f36f73b2d02b72201683cf5795851034bbd6f28e
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Sep 14 12:58:54 2012 -0700

    altosui: Remove debugging from KML export
    
    Left in from testing the Eeprom export bug
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 97ab77d548964115e4b41ad5952194fcd1455c96
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Sep 14 11:13:02 2012 -0700

    altosui: Fix Landed tab units
    
    And clean up the whole flight value reporting code base. It would be
    nice to create a separate class to make this easier; at present
    there's a bunch of customization embedded in how values are presented
    in each tab.
    
    Reported by: Bdale Garbee <bdale@gag.com>
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 43e646657502f6162fa02f37fd2bd5aa3e29a1a8
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Sep 14 11:08:57 2012 -0700

    altoslib: Update GPS seen_values in AltosEepromIterable
    
    Otherwise, KML export won't ever get GPS data to write.
    
    Reported by: Bob Brown <prefect@kloudbusters.org>
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 7b6dd9105ba36aa11d6d0ee6e0823965b9beffb9
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Sep 14 11:08:19 2012 -0700

    altoslib: Remove duplicate seen_ values in AltosEepromIterable
    
    These are defined in AltosRecord; duplicating them is not a good plan
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit eefcfa94f360f8c5a7233370d4178525bccbb22f
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Sep 13 16:13:42 2012 -0600

    add note about checking doc copyright date and revision history to Releasing

commit 2441090f26c3df66a6ce48d64e64384ce9e5ef82
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Sep 13 15:12:33 2012 -0700

    doc: Add revision history for 1.1. Update copyright date
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 320d90c376dccfe1599505e3b485df8d46e34bb3
Merge: 9a7d643 2e6c6a6
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Sep 13 15:36:37 2012 -0600

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos

commit 2e6c6a6c15a7cff9f20b70a4cf58697da1302f01
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Sep 13 00:35:27 2012 -0700

    doc: Document imperial units and groundstation configuration
    
    What it says on the tin.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 9a7d6431777ce3377b788ddac6cb9fadd53c039c
Merge: 2439f53 9728b20
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Sep 13 00:58:30 2012 -0600

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos

commit 9728b20a4ad9576ac1022dc9a5771f5a3eb92b3d
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Sep 12 23:56:18 2012 -0700

    altosui: Exit with an error status when file processing fails
    
    If the user provides any files on the command line, and if processing
    them fails in some way, exit immediately with an error indication.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 0564f9b13d4e3050b0a45fc1c4f710ab2b1f9d81
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Sep 12 22:40:57 2012 -0700

    altosui: Leave new_gps indication until gps value is used
    
    During replay, AltosState may not see a new GPS value as soon as it
    lands in the state field as additional records with the same timestamp
    may come in after the GPS record.
    
    Instead of resetting the new_gps indication when the new record is
    created, wait until the new_gps indication is seen by the AltosState
    update code and have that clear the new_gps indication.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 7e237920abb582d7ceb3e1925b11a848575ac68a
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Sep 12 21:32:25 2012 -0700

    altosui: Stop downloading eeprom data on a block full of invalid data
    
    When no valid records are found within an eeprom block, we assume that
    no more data will be found within the entire storage area.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 2439f53ea21c84f7c510dc191dde025ada24281f
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Sep 12 21:50:25 2012 -0600


ao-tools/ao-usbload/ao-usbload.c
/*
 * Copyright © 2012 Keith Packard <keithp@keithp.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */

#include <err.h>
#include <fcntl.h>
#include <gelf.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <sysexits.h>
#include <unistd.h>
#include <getopt.h>
#include <string.h>
#include <stdbool.h>
#include "ao-elf.h"
#include "ccdbg.h"
#include "cc-usb.h"
#include "cc.h"
#include "ao-usbload.h"
#include "ao-selfload.h"
#include "ao-verbose.h"
#include "ao-editaltos.h"

static uint16_t
get_uint16(struct cc_usb *cc, uint32_t addr)
{
	uint16_t	result;
	result = ao_self_get_uint16(cc, addr);
	printf ("read 0x%08x = 0x%04x\n", addr, result);
	return result;
}

/*
 * Read a 32-bit value from the target device with arbitrary
 * alignment
 */
static uint32_t
get_uint32(struct cc_usb *cc, uint32_t addr)
{
	uint32_t	result;

	result = ao_self_get_uint32(cc, addr);
	printf ("read 0x%08x = 0x%08x\n", addr, result);
	return result;
}

/*
 * Check to see if the target device has been
 * flashed with a similar firmware image before
 *
 * This is done by looking for the same romconfig version,
 * which should be at the same location as the linker script
 * places this at 0x100 from the start of the rom section
 */
static int
check_flashed(struct cc_usb *cc)
{
	uint16_t	romconfig_version = get_uint16(cc, AO_ROMCONFIG_VERSION);
	uint16_t	romconfig_check = get_uint16(cc, AO_ROMCONFIG_CHECK);

	if (romconfig_version != (uint16_t) ~romconfig_check) {
		fprintf (stderr, "Device has not been flashed before\n");
		return 0;
	}
	return 1;
}

static const struct option options[] = {
	{ .name = "tty", .has_arg = 1, .val = 'T' },
	{ .name = "device", .has_arg = 1, .val = 'D' },
	{ .name = "cal", .has_arg = 1, .val = 'c' },
	{ .name = "serial", .has_arg = 1, .val = 's' },
	{ .name = "verbose", .has_arg = 1, .val = 'v' },
	{ 0, 0, 0, 0},
};

static void usage(char *program)
{
	fprintf(stderr, "usage: %s [--verbose=<verbose>] [--device=<device>] [-tty=<tty>] [--cal=<radio-cal>] [--serial=<serial>] file.{elf,ihx}\n", program);
	exit(1);
}

void
done(struct cc_usb *cc, int code)
{
/*	cc_usb_printf(cc, "a\n"); */
	cc_usb_close(cc);
	exit (code);
}

static int
ends_with(char *whole, char *suffix)
{
	int whole_len = strlen(whole);
	int suffix_len = strlen(suffix);

	if (suffix_len > whole_len)
		return 0;
	return strcmp(whole + whole_len - suffix_len, suffix) == 0;
}

int
main (int argc, char **argv)
{
	char			*device = NULL;
	char			*filename;
	Elf			*e;
	char			*serial_end;
	unsigned int		serial = 0;
	char			*serial_ucs2;
	int			serial_ucs2_len;
	char			serial_int[2];
	unsigned int		s;
	int			i;
	int			string_num;
	uint32_t		cal = 0;
	char			cal_int[4];
	char			*cal_end;
	int			c;
	int			was_flashed = 0;
	struct ao_hex_image	*load;
	int			tries;
	struct cc_usb		*cc = NULL;
	char			*tty = NULL;
	int			success;
	int			verbose = 0;
	struct ao_sym		*file_symbols;
	int			num_file_symbols;

	while ((c = getopt_long(argc, argv, "T:D:c:s:v:", options, NULL)) != -1) {
		switch (c) {
		case 'T':
			tty = optarg;
			break;
		case 'D':
			device = optarg;
			break;
		case 'c':
			cal = strtoul(optarg, &cal_end, 10);
			if (cal_end == optarg || *cal_end != '\0')
				usage(argv[0]);
			break;
		case 's':
			serial = strtoul(optarg, &serial_end, 10);
			if (serial_end == optarg || *serial_end != '\0')
				usage(argv[0]);
			break;
		case 'v':
			verbose++;
			break;
		default:
			usage(argv[0]);
			break;
		}
	}

	ao_verbose = verbose;

	if (verbose > 1)
		ccdbg_add_debug(CC_DEBUG_BITBANG);

	filename = argv[optind];
	if (filename == NULL)
		usage(argv[0]);

	if (ends_with (filename, ".elf")) {
		load = ao_load_elf(filename, &file_symbols, &num_file_symbols);
	} else if (ends_with (filename, ".ihx")) {
		load = ao_hex_load(filename, &file_symbols, &num_file_symbols);
	} else
		usage(argv[0]);

	if (ao_editaltos_find_symbols(file_symbols, num_file_symbols, ao_symbols, ao_num_symbols))
		fprintf(stderr, "Cannot find required symbols\n");

	{
		int	is_loader;
		int	tries;

		for (tries = 0; tries < 3; tries++) {
			char	*this_tty = tty;
			if (!this_tty)
				this_tty = cc_usbdevs_find_by_arg(device, "AltosFlash");
			if (!this_tty)
				this_tty = cc_usbdevs_find_by_arg(device, "MegaMetrum");
			if (!this_tty)
				this_tty = getenv("ALTOS_TTY");
			if (!this_tty)
				this_tty="/dev/ttyACM0";

			cc = cc_usb_open(this_tty);

			if (!cc)
				exit(1);
			cc_usb_printf(cc, "v\n");
			is_loader = 0;
			for (;;) {
				char	line[256];
				cc_usb_getline(cc, line, sizeof(line));
				if (!strncmp(line, "altos-loader", 12))
					is_loader = 1;
				if (!strncmp(line, "software-version", 16))
					break;
			}
			if (is_loader)
				break;
			printf ("rebooting to loader\n");
			cc_usb_printf(cc, "X\n");
			cc_usb_close(cc);
			sleep(1);
			cc = NULL;
		}
		if (!is_loader) {
			fprintf(stderr, "Cannot switch to boot loader\n");
			exit(1);
		}
#if 0
		{
			uint8_t	check[256];
			int	i = 0;

			ao_self_block_read(cc, AO_BOOT_APPLICATION_BASE, check);
			for (;;) {
				uint8_t block[256];
				putchar ('.');
				if (++i == 40) {
					putchar('\n');
					i = 0;
				}
				fflush(stdout);
				ao_self_block_write(cc, AO_BOOT_APPLICATION_BASE, block);
				ao_self_block_read(cc, AO_BOOT_APPLICATION_BASE, block);
				if (memcmp(block, check, 256) != 0) {
					fprintf (stderr, "read differed\n");
					exit(1);
				}
			}
		}
#endif
	}

	/* Go fetch existing config values
	 * if available
	 */
	was_flashed = check_flashed(cc);

	if (!serial) {
		if (!was_flashed) {
			fprintf (stderr, "Must provide serial number\n");
			done(cc, 1);
		}
		serial = get_uint16(cc, AO_SERIAL_NUMBER);
		if (!serial || serial == 0xffff) {
			fprintf (stderr, "Invalid existing serial %d\n", serial);
			done(cc, 1);
		}
	}

	if (!cal && AO_RADIO_CAL && was_flashed) {
		cal = get_uint32(cc, AO_RADIO_CAL);
		if (!cal || cal == 0xffffffff) {
			fprintf (stderr, "Invalid existing rf cal %d\n", cal);
			done(cc, 1);
		}
	}

	if (!ao_editaltos(load, serial, cal))
		done(cc, 1);

	/* And flash the resulting image to the device
	 */
	success = ao_self_write(cc, load);
		
	if (!success) {
		fprintf (stderr, "\"%s\": Write failed\n", filename);
		done(cc, 1);
	}

	done(cc, 0);
}
